File: clients/scala/src/main/scala/spb_ai_champ/model/Game.scala
package spb_ai_champ.model

import spb_ai_champ.util.StreamUtil

/**
 * TODO - Document
 *
 * @param myIndex TODO - Document
 * @param currentTick TODO - Document
 * @param maxTickCount TODO - Document
 * @param players TODO - Document
 * @param planets TODO - Document
 * @param flyingWorkerGroups TODO - Document
 * @param maxFlyingWorkerGroups TODO - Document
 * @param maxTravelDistance TODO - Document
 * @param maxBuilders TODO - Document
 * @param buildingProperties TODO - Document
 */
case class Game(myIndex: Int, currentTick: Int, maxTickCount: Int, players: Seq[spb_ai_champ.model.Player], planets: Seq[spb_ai_champ.model.Planet], flyingWorkerGroups: Seq[spb_ai_champ.model.FlyingWorkerGroup], maxFlyingWorkerGroups: Int, maxTravelDistance: Int, maxBuilders: Int, buildingProperties: Map[spb_ai_champ.model.BuildingType, spb_ai_champ.model.BuildingProperties]) {
    /**
     * Write Game to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, myIndex)
        StreamUtil.writeInt(stream, currentTick)
        StreamUtil.writeInt(stream, maxTickCount)
        StreamUtil.writeInt(stream, players.length)
        players.foreach { value =>
            value.writeTo(stream)
        }
        StreamUtil.writeInt(stream, planets.length)
        planets.foreach { value =>
            value.writeTo(stream)
        }
        StreamUtil.writeInt(stream, flyingWorkerGroups.length)
        flyingWorkerGroups.foreach { value =>
            value.writeTo(stream)
        }
        StreamUtil.writeInt(stream, maxFlyingWorkerGroups)
        StreamUtil.writeInt(stream, maxTravelDistance)
        StreamUtil.writeInt(stream, maxBuilders)
        StreamUtil.writeInt(stream, buildingProperties.size)
        buildingProperties.foreach { case (key, value) =>
            key.writeTo(stream)
            value.writeTo(stream)
        }
    }

    /**
     * Get string representation of Game
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Game { ")
        stringBuilder.append("myIndex: ")
        stringBuilder.append(myIndex)
        stringBuilder.append(", ")
        stringBuilder.append("currentTick: ")
        stringBuilder.append(currentTick)
        stringBuilder.append(", ")
        stringBuilder.append("maxTickCount: ")
        stringBuilder.append(maxTickCount)
        stringBuilder.append(", ")
        stringBuilder.append("players: ")
        stringBuilder.append(players)
        stringBuilder.append(", ")
        stringBuilder.append("planets: ")
        stringBuilder.append(planets)
        stringBuilder.append(", ")
        stringBuilder.append("flyingWorkerGroups: ")
        stringBuilder.append(flyingWorkerGroups)
        stringBuilder.append(", ")
        stringBuilder.append("maxFlyingWorkerGroups: ")
        stringBuilder.append(maxFlyingWorkerGroups)
        stringBuilder.append(", ")
        stringBuilder.append("maxTravelDistance: ")
        stringBuilder.append(maxTravelDistance)
        stringBuilder.append(", ")
        stringBuilder.append("maxBuilders: ")
        stringBuilder.append(maxBuilders)
        stringBuilder.append(", ")
        stringBuilder.append("buildingProperties: ")
        stringBuilder.append(buildingProperties)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Game {
    /**
     * Read Game from input stream
     */
    def readFrom(stream: java.io.InputStream): Game = Game(
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        (0 until StreamUtil.readInt(stream)).map { _ =>
            spb_ai_champ.model.Player.readFrom(stream)
        },
        (0 until StreamUtil.readInt(stream)).map { _ =>
            spb_ai_champ.model.Planet.readFrom(stream)
        },
        (0 until StreamUtil.readInt(stream)).map { _ =>
            spb_ai_champ.model.FlyingWorkerGroup.readFrom(stream)
        },
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        (0 until StreamUtil.readInt(stream)).map { _ => (
            spb_ai_champ.model.BuildingType.readFrom(stream),
            spb_ai_champ.model.BuildingProperties.readFrom(stream)
        )}.toMap
    )
}

File: clients/rust/src/model/game.rs
use super::*;

/// TODO - Document
#[derive(Clone, Debug)]
pub struct Game {
    /// TODO - Document
    pub my_index: i32,
    /// TODO - Document
    pub current_tick: i32,
    /// TODO - Document
    pub max_tick_count: i32,
    /// TODO - Document
    pub players: Vec<model::Player>,
    /// TODO - Document
    pub planets: Vec<model::Planet>,
    /// TODO - Document
    pub flying_worker_groups: Vec<model::FlyingWorkerGroup>,
    /// TODO - Document
    pub max_flying_worker_groups: i32,
    /// TODO - Document
    pub max_travel_distance: i32,
    /// TODO - Document
    pub max_builders: i32,
    /// TODO - Document
    pub building_properties: std::collections::HashMap<model::BuildingType, model::BuildingProperties>,
}

impl trans::Trans for Game {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.my_index.write_to(writer)?;
        self.current_tick.write_to(writer)?;
        self.max_tick_count.write_to(writer)?;
        self.players.write_to(writer)?;
        self.planets.write_to(writer)?;
        self.flying_worker_groups.write_to(writer)?;
        self.max_flying_worker_groups.write_to(writer)?;
        self.max_travel_distance.write_to(writer)?;
        self.max_builders.write_to(writer)?;
        self.building_properties.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let my_index: i32 = trans::Trans::read_from(reader)?;
        let current_tick: i32 = trans::Trans::read_from(reader)?;
        let max_tick_count: i32 = trans::Trans::read_from(reader)?;
        let players: Vec<model::Player> = trans::Trans::read_from(reader)?;
        let planets: Vec<model::Planet> = trans::Trans::read_from(reader)?;
        let flying_worker_groups: Vec<model::FlyingWorkerGroup> = trans::Trans::read_from(reader)?;
        let max_flying_worker_groups: i32 = trans::Trans::read_from(reader)?;
        let max_travel_distance: i32 = trans::Trans::read_from(reader)?;
        let max_builders: i32 = trans::Trans::read_from(reader)?;
        let building_properties: std::collections::HashMap<model::BuildingType, model::BuildingProperties> = trans::Trans::read_from(reader)?;
        Ok(Self {
            my_index,
            current_tick,
            max_tick_count,
            players,
            planets,
            flying_worker_groups,
            max_flying_worker_groups,
            max_travel_distance,
            max_builders,
            building_properties,
        })
    }
}